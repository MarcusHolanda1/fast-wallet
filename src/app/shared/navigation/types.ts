import { Card } from '../types/card';

export type RootStackParamList = {
  HomeScreen: undefined;
  RegisterScreen: undefined;
  RegisterSuccessScreen: {
    card: Card;
  };
  WalletScreen: undefined;
};
