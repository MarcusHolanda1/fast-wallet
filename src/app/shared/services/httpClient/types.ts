export interface HttpRequestConfig {
  headers?: Record<string, string>;
  query?: Record<string, string | number | boolean>;
  body?: unknown;
}

export interface HttpResponse<T = unknown> {
  data: T;
  status: number;
}

export class HttpError extends Error {
  public status: number;
  public data: unknown;

  constructor(message: string, status: number, data?: unknown) {
    super(message);
    this.name = 'HttpError';
    this.status = status;
    this.data = data;
  }
}


export interface HttpClient {
  get<T = unknown>(
    url: string,
    config?: HttpRequestConfig
  ): Promise<HttpResponse<T>>;
  post<T = unknown>(
    url: string,
    config?: HttpRequestConfig
  ): Promise<HttpResponse<T>>;
  put<T = unknown>(
    url: string,
    config?: HttpRequestConfig
  ): Promise<HttpResponse<T>>;
  patch<T = unknown>(
    url: string,
    config?: HttpRequestConfig
  ): Promise<HttpResponse<T>>;
  delete<T = unknown>(
    url: string,
    config?: HttpRequestConfig
  ): Promise<HttpResponse<T>>;
}
