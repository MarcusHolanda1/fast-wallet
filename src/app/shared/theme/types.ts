export interface ColorPalette {
  text: {
    black: string;
    grey: string;
    greyDark: string;
  };
  base: {
    white: string;
    greenLight: string;
    blueDark: string;
    blueLight: string;
    greyLight: string;
  };
  alert: {
    yellow: string;
    red: string;
    green: string;
  };
}

export interface FontStyle {
  fontSize: number;
  lineHeight: number;
  fontWeight?: '400' | '500' | '600' | '700';
}

export interface Typography {
  h1: FontStyle;
  h2: FontStyle;
  h3: FontStyle;
  h4: FontStyle;
  h5: FontStyle;
  p: FontStyle;
  pSmall: FontStyle;
}

export interface AppTheme {
  colors: ColorPalette;
  typography: Typography;
}
