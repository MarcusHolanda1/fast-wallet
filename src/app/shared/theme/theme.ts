import { colors } from './colors';
import { typography } from './typography';
import { AppTheme } from './types';

export const theme: AppTheme = {
  colors,
  typography
};

export type { AppTheme };
