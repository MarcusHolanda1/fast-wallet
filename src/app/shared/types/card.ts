export interface Card {
  id: string;
  number: string;
  cvv: string;
  name: string;
  expires: string;
}

export interface CardPayload {
  number: string;
  cvv: string;
  name: string;
  expires: string;
}
