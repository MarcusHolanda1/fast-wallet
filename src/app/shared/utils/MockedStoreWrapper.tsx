import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';

const createTestStore = () => {
  return configureStore({
    reducer: {
      test: (state: Record<string, unknown> = {}) => state
    }
  });
};

export const TestWrapper = ({ children }: { children: React.ReactNode }) => {
  const store = createTestStore();
  return <Provider store={store}>{children}</Provider>;
};
