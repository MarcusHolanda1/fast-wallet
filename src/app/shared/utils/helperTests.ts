import { fireEvent, RenderResult } from '@testing-library/react-native';

export const fillAndBlur = (
  sut: RenderResult,
  testId: string,
  value: string
) => {
  fireEvent.changeText(sut.getByTestId(testId), value);
  fireEvent(sut.getByTestId(testId), 'blur');
};

export const expectFormattedCardNumber = (cardNumber: string) => {
  return '•••• •••• •••• ' + cardNumber.slice(-4);
};
