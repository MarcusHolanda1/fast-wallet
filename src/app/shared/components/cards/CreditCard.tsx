import { View, Text, StyleSheet } from 'react-native';
import { theme } from '@shared/theme/theme';
import { Card } from '@app/shared/types/card';

interface CreditCardProps {
  card: Card;
  backgroundColor?: string;
  textColor?: string;
}

export default function CreditCard({
  card,
  backgroundColor,
  textColor
}: CreditCardProps) {
  return (
    <View style={[styles.cardContainer, { backgroundColor: backgroundColor }]}>
      <Text style={[styles.cardType, { color: textColor }]}>{card.id}</Text>

      <View style={styles.cardContent}>
        <Text style={[styles.cardholderName, { color: textColor }]}>
          {card.name}
        </Text>
        <Text style={[styles.cardNumber, { color: textColor }]}>
          {card.number}
        </Text>
        <Text style={[styles.expiryDate, { color: textColor }]}>
          Validade {card.expires}
        </Text>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  cardContainer: {
    height: 200,
    width: '100%',
    borderRadius: 16,
    paddingHorizontal: 16,
    paddingTop: 30,
    paddingBottom: 40,
    justifyContent: 'space-between',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 4
    },
    shadowOpacity: 0.3,
    shadowRadius: 8,
    elevation: 8
  },
  cardType: {
    ...theme.typography.h5,
    marginBottom: 3
  },
  cardContent: {
    flex: 1,
    justifyContent: 'flex-end'
  },
  cardholderName: {
    ...theme.typography.p,
    marginBottom: 3
  },
  cardNumber: {
    ...theme.typography.pSmall,
    marginBottom: 3
  },
  expiryDate: {
    ...theme.typography.pSmall
  }
});
