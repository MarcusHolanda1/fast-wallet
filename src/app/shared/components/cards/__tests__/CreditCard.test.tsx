import { render, RenderResult } from '@testing-library/react-native';
import { TestWrapper } from '@app/shared/utils/MockedStoreWrapper';
import { Card } from '@app/shared/types/card';
import { expectFormattedCardNumber } from '@app/shared/utils/helperTests';
import { theme } from '@app/shared/theme/theme';

import { generateFakeCardData } from '../../../../../__mocks__/card';
import CreditCard from '../CreditCard';

const mockFakeCard = generateFakeCardData();

const backgroundColor = theme.colors.base.greenLight;
const textColor = '#fff';

const makeSut = (cardProps: Card) => {
  return render(
    <TestWrapper>
      <CreditCard
        card={cardProps}
        backgroundColor={backgroundColor}
        textColor={textColor}
        testID={`credit-card-${cardProps.id}`}
      />
    </TestWrapper>
  );
};

describe('Credit card', () => {
  let sut: RenderResult;

  beforeEach(() => {
    jest.clearAllMocks();
    sut = makeSut(mockFakeCard);
  });

  test('should render the credit card with correct details', () => {
    expect(sut.getByText('Green')).toBeVisible();
    expect(
      sut.getByText(expectFormattedCardNumber(mockFakeCard.number))
    ).toBeTruthy();
    expect(sut.getByText(mockFakeCard.name)).toBeTruthy();
    expect(sut.getByText(`Validade ${mockFakeCard.expires}`)).toBeTruthy();
  });

  test('should apply background and text colors', () => {
    expect(sut.getByTestId(`credit-card-${mockFakeCard.id}`)).toHaveStyle({
      backgroundColor
    });

    expect(sut.getByText('Green')).toHaveStyle({ color: textColor });
    expect(sut.getByText(mockFakeCard.name)).toHaveStyle({
      color: textColor
    });
    expect(
      sut.getByText(expectFormattedCardNumber(mockFakeCard.number))
    ).toHaveStyle({
      color: textColor
    });
    expect(sut.getByText(`Validade ${mockFakeCard.expires}`)).toHaveStyle({
      color: textColor
    });
  });

  test('should match snapshot', () => {
    expect(sut.toJSON()).toMatchSnapshot();
  });
});
