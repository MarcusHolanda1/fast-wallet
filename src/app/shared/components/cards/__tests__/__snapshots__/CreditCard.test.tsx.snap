// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Credit card should match snapshot 1`] = `
<View
  style={
    [
      {
        "borderRadius": 16,
        "elevation": 8,
        "height": 200,
        "justifyContent": "space-between",
        "paddingBottom": 40,
        "paddingHorizontal": 16,
        "paddingTop": 30,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 8,
        "width": "100%",
      },
      {
        "backgroundColor": "#000",
      },
    ]
  }
  testID="credit-card-1"
>
  <Text
    style={
      [
        {
          "fontSize": 18,
          "fontWeight": "400",
          "lineHeight": 20,
          "marginBottom": 3,
        },
        {
          "color": "#fff",
        },
      ]
    }
  >
    1
  </Text>
  <View
    style={
      {
        "flex": 1,
        "justifyContent": "flex-end",
      }
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 18,
            "marginBottom": 3,
          },
          {
            "color": "#fff",
          },
        ]
      }
    >
      MARCUS HOLANDA
    </Text>
    <Text
      accessibilityLabel="Final 1111"
      style={
        [
          {
            "color": "#fff",
          },
          {
            "fontSize": 14,
            "fontWeight": "400",
            "letterSpacing": 1,
            "lineHeight": 16,
            "marginBottom": 3,
          },
        ]
      }
      testID="credit-card-1-number"
    >
      •••• •••• •••• 1111
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 14,
            "fontWeight": "400",
            "lineHeight": 16,
          },
          {
            "color": "#fff",
          },
        ]
      }
    >
      Validade 
      12/30
    </Text>
  </View>
</View>
`;
